src/java.base/share/classes/sun/security/rsa/RSASignature.java

```diff
@@ -205,27 +207,27 @@ protected boolean engineVerify(byte[] sigBytes) throws SignatureException {
         }
         try {
             if (sigBytes.length != RSACore.getByteLength(publicKey)) {
-                throw new SignatureException("Signature length not correct: got " +
+                throw new SignatureException("Bad signature length: got " +
                     sigBytes.length + " but was expecting " +
                     RSACore.getByteLength(publicKey));
             }
-            byte[] digest = getDigestValue();
-            byte[] decrypted = RSACore.rsa(sigBytes, publicKey);
-            byte[] unpadded = padding.unpad(decrypted);
+
             // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
-            // Step 4 suggests comparing the encoded message instead of the
-            // decoded, but some vendors might omit the NULL params in
-            // digest algorithm identifier.
-            byte[] decodedDigest = RSAUtil.decodeSignature(digestOID, unpadded);
-            return MessageDigest.isEqual(digest, decodedDigest);
+            // Step 4 suggests comparing the encoded message
+            byte[] decrypted = RSACore.rsa(sigBytes, publicKey);
+
+            byte[] digest = getDigestValue();
+            byte[] encoded = RSAUtil.encodeSignature(digestOID, digest);
+            byte[] padded = padding.pad(encoded);
+            if (!MessageDigest.isEqual(padded, decrypted)) {
+                byte[] encoded2 = RSAUtil.encodeSignatureOmitNull
+                        (digestOID, digest);
+                byte[] padded2 = padding.pad(encoded2);
+                return MessageDigest.isEqual(padded2, decrypted);
+            }
```

Comment:
I had a check of the specification (Section A.2.4 of RFC 8017), and the [update](https://github.com/openjdk/jdk/pull/8365) and the [JBS entry](https://bugs.openjdk.org/browse/JDK-8285404) that added the comment "some vendors might omit the NULL params".

Per section A.2.4 of RFC 8017, it is said "For each OID, the parameters field associated with this OID in a value of type  AlgorithmIdentifier SHALL have a value of type NULL."

Per the key words specification, RFC 2119, "SHALL" is the same as MUST which  "mean that the definition is an absolute requirement of the specification."

In the bug description of bug JDK-8285404, there is a section "*Update*: We think it's possible that there might be signers omitting the NULL params in the digest algorithm identifier. "

For this case, if the signers omitting the NULL params, does it means the signer does not follow the specification and should be rejected?  @wangweij could you recall if there is a real case that omits the NULL params in practice?

Reply:
Max is on vacation and may not see your question for a while...
IIRC, the inconsistency (NULL vs omission) goes way back. As time goes on, this may no longer be an issue as spec is clarified and vendors update their implementation.

Reply:
I checked back the specification back to RFC 2437, released on October 1998, which requires to encode NULL parameters as well.  As the update to keep the consistency is not trivial, I may just remove it and see if it could be a real problem in practice.

Reply:
Yes, let's try it and see.

Reply:
Thank you.  I have no more comments.